File: cli/node/src/commands/entities.ts
/**
 * Entity management commands.
 */

import readline from "node:readline";
import Table from "cli-table3";
import type { Backend } from "../backend/base.js";
import {
	colors,
	printError,
	printInfo,
	printSuccess,
	timedStatus,
} from "../branding.js";
import { formatAgentEnvelope, formatJson } from "../output.js";
import { setCurrentCommand } from "../state.js";

const { brand, accent, dim } = colors;

const VALID_TYPES = new Set(["users", "agents", "apps", "runs"]);

export async function cmdEntitiesList(
	backend: Backend,
	entityType: string,
	opts: { output: string },
): Promise<void> {
	setCurrentCommand("entity list");
	if (!VALID_TYPES.has(entityType)) {
		printError(
			`Invalid entity type: ${entityType}. Use: ${[...VALID_TYPES].join(", ")}`,
		);
		process.exit(1);
	}

	const start = performance.now();
	let results: Record<string, unknown>[];
	try {
		results = await timedStatus(`Fetching ${entityType}...`, async () => {
			return backend.entities(entityType);
		});
	} catch (e) {
		printError(
			e instanceof Error ? e.message : String(e),
			"This feature may require the mem0 Platform.",
		);
		process.exit(1);
	}
	const elapsed = (performance.now() - start) / 1000;

	if (opts.output === "agent" || opts.output === "json") {
		formatAgentEnvelope({
			command: "entity list",
			data: results,
			count: results.length,
			durationMs: Math.round(elapsed * 1000),
		});
		return;
	}

	if (!results.length) {
		printInfo(`No ${entityType} found.`);
		return;
	}

	const table = new Table({
		head: [accent("Name / ID"), accent("Created")],
		style: { head: [], border: [] },
	});

	for (const entity of results) {
		const name = String(entity.name ?? entity.id ?? "—");
		const created = String(entity.created_at ?? "—").slice(0, 10);
		table.push([name, created]);
	}

	console.log();
	console.log(table.toString());
	console.log(
		`  ${dim(`${results.length} ${entityType} (${elapsed.toFixed(2)}s)`)}`,
	);
	console.log();
}

export async function cmdEntitiesDelete(
	backend: Backend,
	opts: {
		userId?: string;
		agentId?: string;
		appId?: string;
		runId?: string;
		dryRun?: boolean;
		force: boolean;
		output: string;
	},
): Promise<void> {
	setCurrentCommand("entity delete");
	const { isAgentMode } = await import("../state.js");
	if (isAgentMode() && !opts.force) {
		printError("Destructive operation requires --force in agent mode.");
		process.exit(1);
	}
	if (!opts.userId && !opts.agentId && !opts.appId && !opts.runId) {
		printError(
			"Provide at least one of --user-id, --agent-id, --app-id, --run-id.",
		);
		process.exit(1);
	}

	const scopeParts: string[] = [];
	if (opts.userId) scopeParts.push(`user=${opts.userId}`);
	if (opts.agentId) scopeParts.push(`agent=${opts.agentId}`);
	if (opts.appId) scopeParts.push(`app=${opts.appId}`);
	if (opts.runId) scopeParts.push(`run=${opts.runId}`);
	const scope = scopeParts.join(", ");

	if (opts.dryRun) {
		printInfo(`Would delete entity ${scope} and all its memories.`);
		printInfo("No changes made.");
		return;
	}

	if (!opts.force) {
		const rl = readline.createInterface({
			input: process.stdin,
			output: process.stdout,
		});
		const answer = await new Promise<string>((resolve) => {
			rl.question(
				`\n  \u26a0  Delete entity ${scope} AND all its memories? This cannot be undone. [y/N] `,
				resolve,
			);
		});
		rl.close();
		if (answer.toLowerCase() !== "y") {
			printInfo("Cancelled.");
			process.exit(0);
		}
	}

	const start = performance.now();
	let result: Record<string, unknown>;
	try {
		result = await timedStatus("Deleting entity...", async () => {
			return backend.deleteEntities({
				userId: opts.userId,
				agentId: opts.agentId,
				appId: opts.appId,
				runId: opts.runId,
			});
		});
	} catch (e) {
		printError(e instanceof Error ? e.message : String(e));
		process.exit(1);
	}
	const elapsed = (performance.now() - start) / 1000;

	if (opts.output === "agent") {
		formatAgentEnvelope({
			command: "entity delete",
			data: { deleted: true },
			durationMs: Math.round(elapsed * 1000),
		});
	} else if (opts.output === "json") {
		formatJson(result);
	} else if (opts.output !== "quiet") {
		printSuccess(`Entity deleted with all memories (${elapsed.toFixed(2)}s)`);
	}
}

File: cli/python/src/mem0_cli/commands/entities.py
"""Entity management commands."""

from __future__ import annotations

import time as _time

import typer
from rich.console import Console
from rich.table import Table

from mem0_cli.backend.base import Backend
from mem0_cli.branding import (
    ACCENT_COLOR,
    BRAND_COLOR,
    DIM_COLOR,
    print_error,
    print_info,
    print_success,
    timed_status,
)
from mem0_cli.output import format_agent_envelope, format_json

console = Console()
err_console = Console(stderr=True)


def cmd_entities_list(backend: Backend, entity_type: str, *, output: str) -> None:
    """List entities of a given type."""
    from mem0_cli.state import is_agent_mode, set_current_command

    set_current_command("entity list")
    if is_agent_mode():
        output = "agent"
    valid_types = {"users", "agents", "apps", "runs"}
    if entity_type not in valid_types:
        print_error(
            err_console, f"Invalid entity type: {entity_type}. Use: {', '.join(valid_types)}"
        )
        raise typer.Exit(1)

    _start = _time.perf_counter()
    with timed_status(err_console, f"Fetching {entity_type}...") as _ts:
        try:
            results = backend.entities(entity_type)
        except Exception as e:
            print_error(err_console, str(e), hint="This feature may require the mem0 Platform.")
            raise typer.Exit(1) from None
    _elapsed = _time.perf_counter() - _start

    if output == "agent":
        format_agent_envelope(
            console,
            command="entity list",
            data=results,
            count=len(results),
            duration_ms=int(_elapsed * 1000),
        )
        return

    if output == "json":
        format_json(console, results)
        return

    if not results:
        print_info(console, f"No {entity_type} found.")
        return

    table = Table(border_style=BRAND_COLOR, header_style=f"bold {ACCENT_COLOR}", padding=(0, 1))
    table.add_column("Name / ID", style="bold")
    table.add_column("Created", max_width=12)

    for entity in results:
        name = entity.get("name", entity.get("id", "—"))
        created = str(entity.get("created_at", "—"))[:10]
        table.add_row(str(name), created)

    console.print()
    console.print(table)
    console.print(f"  [{DIM_COLOR}]{len(results)} {entity_type} ({_elapsed:.2f}s)[/]")
    console.print()


def cmd_entities_delete(
    backend: Backend,
    *,
    user_id: str | None,
    agent_id: str | None,
    app_id: str | None,
    run_id: str | None,
    force: bool,
    dry_run: bool = False,
    output: str,
) -> None:
    """Delete an entity and all its memories (cascade delete)."""
    from mem0_cli.state import is_agent_mode, set_current_command

    set_current_command("entity delete")
    if is_agent_mode():
        output = "agent"
        if not force:
            print_error(err_console, "Destructive operation requires --force in agent mode.")
            raise typer.Exit(1)
    if not any([user_id, agent_id, app_id, run_id]):
        print_error(
            err_console, "Provide at least one of --user-id, --agent-id, --app-id, --run-id."
        )
        raise typer.Exit(1)

    scope_parts = []
    if user_id:
        scope_parts.append(f"user={user_id}")
    if agent_id:
        scope_parts.append(f"agent={agent_id}")
    if app_id:
        scope_parts.append(f"app={app_id}")
    if run_id:
        scope_parts.append(f"run={run_id}")
    scope_str = ", ".join(scope_parts)

    if dry_run:
        print_info(console, f"Would delete entity {scope_str} and all its memories.")
        print_info(console, "No changes made (dry run).")
        return

    if not force:
        confirm = typer.confirm(
            f"\n  \u26a0  Delete entity {scope_str} AND all its memories? This cannot be undone."
        )
        if not confirm:
            print_info(console, "Cancelled.")
            raise typer.Exit(0)

    _start = _time.perf_counter()
    with timed_status(err_console, "Deleting entity...") as _ts:
        try:
            result = backend.delete_entities(
                user_id=user_id,
                agent_id=agent_id,
                app_id=app_id,
                run_id=run_id,
            )
        except Exception as e:
            print_error(err_console, str(e))
            raise typer.Exit(1) from None
    _elapsed = _time.perf_counter() - _start

    scope = {
        k: v
        for k, v in {
            "user_id": user_id,
            "agent_id": agent_id,
            "app_id": app_id,
            "run_id": run_id,
        }.items()
        if v
    }
    if output == "agent":
        format_agent_envelope(
            console,
            command="entity delete",
            data={"deleted": True},
            scope=scope or None,
            duration_ms=int(_elapsed * 1000),
        )
    elif output == "json":
        format_json(console, result)
    elif output != "quiet":
        print_success(console, f"Entity deleted with all memories ({_elapsed:.2f}s)")
